PSReadLine/Render.cs

```diff
@@ -310,38 +311,61 @@ void MaybeEmphasize(int i, string currColor)
                     activeColor = string.Empty;
 
                     UpdateColorsIfNecessary(Options._continuationPromptColor);
-                    foreach (char c in Options.ContinuationPrompt)
-                    {
-                        _consoleBufferLines[currentLogicalLine].Append(c);
-                    }
+                    _consoleBufferLines[currentLogicalLine].Append(Options.ContinuationPrompt);
 
                     if (inSelectedRegion)
                     {
                         // Turn off inverse before end of line, turn on after continuation prompt
                         _consoleBufferLines[currentLogicalLine].Append(Options.SelectionColor);
                     }
                 }
+                else if (char.IsControl(charToRender))
+                {
+                    MaybeEmphasize(i, color);
+                    _consoleBufferLines[currentLogicalLine].Append('^');
+                    _consoleBufferLines[currentLogicalLine].Append((char)('@' + charToRender));
+                }
                 else
                 {
-                    if (char.IsControl(charToRender))
+                    MaybeEmphasize(i, color);
+                    _consoleBufferLines[currentLogicalLine].Append(charToRender);
+                }
+            }
+
+            if (inSelectedRegion)
+            {
+                _consoleBufferLines[currentLogicalLine].Append("\x1b[0m");
+            }
+
+            inSelectedRegion = false;
+            if (suggestion != null)
+            {
```

Comment:
I'm slightly worried about duplicated logic in this loop and the main rendering loop (which already has special code to cover text after the last token for cases where the parser fails to produce enough tokens).

At least consider refactoring the inner loop into a function that could be shared.

Reply:
Very valid concern. I have refactored the code to render one char into a function.